File: backend/app/schemas/commissioning.py
"""
Pydantic schemas for P5 HV Commissioning Simulation.

Request and response models for the equipment state machine,
switching programme execution, LOTO tracking, and PiC decisions.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

# ── Equipment Schemas ────────────────────────────────────────────


class EquipmentStateSchema(BaseModel):
    """Current state of a single piece of equipment."""

    equipment_id: str = Field(description="Equipment identifier")
    equipment_type: str = Field(description="Equipment type (circuit_breaker, etc.)")
    voltage_kv: float = Field(description="Rated voltage in kV")
    location: str = Field(description="Physical location")
    state: str = Field(description="Current state (open, closed, earthed, etc.)")


class InterlockViolationSchema(BaseModel):
    """A safety interlock violation."""

    interlock_id: str = Field(description="Interlock code (e.g. ILK-001)")
    description: str = Field(description="Human-readable explanation")
    blocking_equipment: str = Field(description="Equipment causing the block")
    blocking_state: str = Field(description="State of the blocking equipment")


# ── Programme Schemas ────────────────────────────────────────────


class CreateProgrammeRequest(BaseModel):
    """Request to create a new switching programme."""

    pic_name: str = Field(
        min_length=1,
        description="Person in Control (PiC) name",
    )


class StepSchema(BaseModel):
    """A single step in the switching programme."""

    step_id: str = Field(description="Step identifier (e.g. S-001)")
    step_number: int = Field(description="Sequential position")
    phase: int = Field(description="Programme phase (1-3)")
    step_type: str = Field(description="Step type (check, switching, etc.)")
    action: str = Field(description="Action description")
    equipment_id: str = Field(description="Target equipment (empty if N/A)")
    responsible: str = Field(description="Who performs (PiC, Local, SCADA)")
    pic_confirmation: bool = Field(description="PiC confirmation required")
    verification: str = Field(description="How to verify success")
    notes: str = Field(description="Safety notes")
    status: str = Field(description="Execution status")
    executed_at: datetime | None = Field(default=None, description="Execution timestamp")
    executed_by: str = Field(default="", description="Who executed")


class ProgrammeSummarySchema(BaseModel):
    """Summary of a switching programme for list responses."""

    programme_id: str = Field(description="Programme identifier")
    title: str = Field(description="Programme title")
    pic_name: str = Field(description="Person in Control")
    status: str = Field(description="Programme lifecycle state")
    total_steps: int = Field(description="Total number of steps")
    completed_steps: int = Field(description="Steps completed so far")
    current_step_index: int = Field(description="Next step index (0-based)")
    created_at: datetime = Field(description="Creation timestamp")


class ProgrammeDetailSchema(BaseModel):
    """Detailed programme view with steps and system state."""

    programme_id: str = Field(description="Programme identifier")
    title: str = Field(description="Programme title")
    pic_name: str = Field(description="Person in Control")
    status: str = Field(description="Programme lifecycle state")
    steps: list[StepSchema] = Field(description="All programme steps")
    current_step_index: int = Field(description="Next step index (0-based)")
    equipment_states: list[EquipmentStateSchema] = Field(
        description="Current equipment states",
    )
    created_at: datetime = Field(description="Creation timestamp")


# ── Execution Schemas ────────────────────────────────────────────


class ExecuteStepRequest(BaseModel):
    """Request to execute a step."""

    executed_by: str = Field(
        min_length=1,
        description="Person executing the step",
    )
    pic_confirmed: bool = Field(
        default=True,
        description="PiC has verbally confirmed this step",
    )


class ExecuteStepResponse(BaseModel):
    """Response after executing a step."""

    success: bool = Field(description="True if step completed")
    step_id: str = Field(description="Step that was executed")
    status: str = Field(description="Step status after execution")
    message: str = Field(description="Human-readable result")
    programme_status: str = Field(description="Programme status after step")


class PiCDecisionRequest(BaseModel):
    """Request for PiC GO/NO-GO decision at a hold point."""

    pic_name: str = Field(
        min_length=1,
        description="PiC making the decision",
    )
    decision: str = Field(
        description="Decision: 'go' or 'nogo'",
        pattern="^(go|nogo)$",
    )
    reason: str = Field(
        default="",
        description="Reason (mandatory for NO-GO)",
    )


class PiCDecisionResponse(BaseModel):
    """Response after PiC decision."""

    decision: str = Field(description="Decision made (go/nogo)")
    programme_status: str = Field(description="Programme status after decision")
    message: str = Field(description="Human-readable result")


class EmergencyStopRequest(BaseModel):
    """Request for emergency stop."""

    initiated_by: str = Field(
        min_length=1,
        description="Person initiating emergency stop",
    )
    reason: str = Field(
        min_length=1,
        description="Reason for emergency stop",
    )


class EmergencyStopResponse(BaseModel):
    """Response after emergency stop."""

    success: bool = Field(description="True if stop was executed")
    programme_status: str = Field(description="Programme status (aborted)")
    message: str = Field(description="Human-readable result")


# ── LOTO Schemas ─────────────────────────────────────────────────


class LOTOPointSchema(BaseModel):
    """A single LOTO isolation point."""

    point_id: str = Field(description="LOTO point identifier")
    equipment_id: str = Field(description="Associated earth switch")
    status: str = Field(description="LOTO status (not_applied, applied, removed)")
    locked_by: str = Field(default="", description="Who applied the lock")
    tag_number: str = Field(description="Danger tag number")
    applied_at: datetime | None = Field(default=None, description="Lock application time")
    removed_at: datetime | None = Field(default=None, description="Lock removal time")
    removed_by: str = Field(default="", description="Who removed the lock")


class LOTOSetSchema(BaseModel):
    """LOTO set for a programme."""

    programme_id: str = Field(description="Associated programme")
    points: list[LOTOPointSchema] = Field(description="All isolation points")
    all_applied: bool = Field(description="True if all points are APPLIED")
    all_removed: bool = Field(description="True if all points are REMOVED")


class LOTOActionRequest(BaseModel):
    """Request to apply or remove LOTO."""

    performed_by: str = Field(
        min_length=1,
        description="Person performing the action",
    )


class LOTOActionResponse(BaseModel):
    """Response after LOTO action."""

    success: bool = Field(description="True if action succeeded")
    point_id: str = Field(description="Affected LOTO point")
    status: str = Field(description="New LOTO status")
    message: str = Field(description="Human-readable result")


# ── Audit Trail Schema ──────────────────────────────────────────


class AuditRecordSchema(BaseModel):
    """A single audit trail entry."""

    record_id: str = Field(description="Record identifier")
    timestamp: datetime = Field(description="UTC timestamp")
    action: str = Field(description="Action description")
    performed_by: str = Field(description="Who performed")
    step_id: str = Field(default="", description="Associated step")
    details: str = Field(default="", description="Additional context")


class AuditTrailResponse(BaseModel):
    """Complete audit trail for a programme."""

    programme_id: str = Field(description="Programme identifier")
    total_records: int = Field(description="Total audit entries")
    records: list[AuditRecordSchema] = Field(description="All audit records")

File: frontend/src/types/commissioning.ts
/**
 * TypeScript interfaces mirroring backend Pydantic schemas.
 *
 * All field names use snake_case to match the API JSON directly — no
 * transform layer needed. This simplifies debugging: the shape you see
 * in DevTools is the shape you use in code.
 *
 * Source of truth: backend/app/schemas/commissioning.py
 */

// ── Enums (string unions matching backend StrEnum values) ──────

export type EquipmentType =
  | "circuit_breaker"
  | "disconnector"
  | "earth_switch"
  | "transformer";

export type EquipmentStateValue =
  | "open"
  | "closed"
  | "earthed"
  | "racked_in"
  | "racked_out";

export type ProgrammeStatus =
  | "created"
  | "approved"
  | "in_progress"
  | "hold"
  | "completed"
  | "aborted";

export type StepStatus =
  | "pending"
  | "in_progress"
  | "completed"
  | "skipped"
  | "hold_point";

export type StepType =
  | "check"
  | "switching"
  | "verification"
  | "hold_point"
  | "communication";

export type LOTOStatus = "not_applied" | "applied" | "removed";

export type TestVerdict = "pass" | "fail" | "not_tested" | "conditional_pass";

export type CampaignStatus = "in_progress" | "completed" | "approved";

export type PiCDecision = "go" | "nogo";

// ── Equipment Schemas ──────────────────────────────────────────

export interface EquipmentState {
  equipment_id: string;
  equipment_type: EquipmentType;
  voltage_kv: number;
  location: string;
  state: EquipmentStateValue;
}

export interface InterlockViolation {
  interlock_id: string;
  description: string;
  blocking_equipment: string;
  blocking_state: string;
}

// ── Programme Schemas ──────────────────────────────────────────

export interface Step {
  step_id: string;
  step_number: number;
  phase: number;
  step_type: StepType;
  action: string;
  equipment_id: string;
  responsible: string;
  pic_confirmation: boolean;
  verification: string;
  notes: string;
  status: StepStatus;
  executed_at: string | null;
  executed_by: string;
}

export interface ProgrammeSummary {
  programme_id: string;
  title: string;
  pic_name: string;
  status: ProgrammeStatus;
  total_steps: number;
  completed_steps: number;
  current_step_index: number;
  created_at: string;
}

export interface ProgrammeDetail {
  programme_id: string;
  title: string;
  pic_name: string;
  status: ProgrammeStatus;
  steps: Step[];
  current_step_index: number;
  equipment_states: EquipmentState[];
  created_at: string;
}

// ── Execution Schemas ──────────────────────────────────────────

export interface ExecuteStepRequest {
  executed_by: string;
  pic_confirmed: boolean;
}

export interface ExecuteStepResponse {
  success: boolean;
  step_id: string;
  status: string;
  message: string;
  programme_status: ProgrammeStatus;
}

export interface PiCDecisionRequest {
  pic_name: string;
  decision: PiCDecision;
  reason: string;
}

export interface PiCDecisionResponse {
  decision: PiCDecision;
  programme_status: ProgrammeStatus;
  message: string;
}

export interface EmergencyStopRequest {
  initiated_by: string;
  reason: string;
}

export interface EmergencyStopResponse {
  success: boolean;
  programme_status: ProgrammeStatus;
  message: string;
}

// ── LOTO Schemas ───────────────────────────────────────────────

export interface LOTOPoint {
  point_id: string;
  equipment_id: string;
  status: LOTOStatus;
  locked_by: string;
  tag_number: string;
  applied_at: string | null;
  removed_at: string | null;
  removed_by: string;
}

export interface LOTOSet {
  programme_id: string;
  points: LOTOPoint[];
  all_applied: boolean;
  all_removed: boolean;
}

export interface LOTOActionRequest {
  performed_by: string;
}

export interface LOTOActionResponse {
  success: boolean;
  point_id: string;
  status: LOTOStatus;
  message: string;
}

// ── Audit Trail ────────────────────────────────────────────────

export interface AuditRecord {
  record_id: string;
  timestamp: string;
  action: string;
  performed_by: string;
  step_id: string;
  details: string;
}

export interface AuditTrailResponse {
  programme_id: string;
  total_records: number;
  records: AuditRecord[];
}

// ── FAT / SAT Schemas ──────────────────────────────────────────

export interface TestSpecification {
  test_id: string;
  name: string;
  standard: string;
  description: string;
  unit: string;
  min_value: number;
  max_value: number;
}

export interface TestResult {
  test_id: string;
  measured_value: number;
  verdict: TestVerdict;
  recorded_by: string;
  recorded_at: string;
  notes: string;
}

export interface FATCampaign {
  campaign_id: string;
  equipment_tag: string;
  status: CampaignStatus;
  specs: TestSpecification[];
  results: TestResult[];
  all_passed: boolean;
  created_at: string;
  approved_by: string;
  approved_at: string | null;
}

export interface SATCampaign {
  campaign_id: string;
  programme_id: string;
  status: CampaignStatus;
  fat_campaign_id: string;
  specs: TestSpecification[];
  results: TestResult[];
  all_passed: boolean;
  created_at: string;
  approved_by: string;
  approved_at: string | null;
}

// ── Protection Relay Schemas ───────────────────────────────────

export interface RelaySetting {
  setting_id: string;
  function: string;
  description: string;
  pickup_value: number;
  pickup_unit: string;
  time_delay: number;
  location: string;
  standard: string;
}

export interface GradingPair {
  pair_id: string;
  downstream_id: string;
  upstream_id: string;
  required_margin_ms: number;
  description: string;
}

export interface GradingResult {
  pair_id: string;
  downstream_id: string;
  upstream_id: string;
  downstream_delay_s: number;
  upstream_delay_s: number;
  actual_margin_ms: number;
  required_margin_ms: number;
  verdict: "selective" | "non_selective";
}

export interface ProtectionCoordination {
  settings: RelaySetting[];
  grading_pairs: GradingPair[];
  results: GradingResult[];
  all_selective: boolean;
}

// ── Emergency Response Schemas ─────────────────────────────────

export type EmergencyType =
  | "arc_flash"
  | "sf6_leak"
  | "medical"
  | "man_overboard"
  | "comms_failure"
  | "unexpected_voltage";

export type SeverityLevel = "critical" | "high" | "medium";

export interface EmergencyProcedure {
  emergency_type: EmergencyType;
  severity: SeverityLevel;
  immediate_actions: string[];
  responsible: string;
  reference_document: string;
  automated_scada_actions: string[];
  communication_protocol: string[];
}

export interface EmergencyEvent {
  event_id: string;
  programme_id: string;
  emergency_type: EmergencyType;
  severity: SeverityLevel;
  triggered_by: string;
  triggered_at: string;
  actions_taken: string[];
  scada_actions_executed: string[];
  resolved: boolean;
  resolved_at: string | null;
}

export interface EmergencyLogResponse {
  programme_id: string;
  total_events: number;
  events: EmergencyEvent[];
}

// ── Grid Code Compliance Schemas ───────────────────────────────

export type NotificationStage = "eon" | "ion" | "fon";

export type ComplianceVerdict =
  | "compliant"
  | "non_compliant"
  | "pending"
  | "conditional";

export interface GridCodeTest {
  test_id: string;
  stage: NotificationStage;
  name: string;
  description: string;
  standard: string;
  acceptance_criteria: string;
  verdict: ComplianceVerdict;
  evidence: string;
  tested_by: string;
  tested_at: string | null;
}

export interface NotificationApplication {
  stage: NotificationStage;
  status: ComplianceVerdict;
  tests: GridCodeTest[];
  submitted_to: string;
  submitted_at: string | null;
  approved_at: string | null;
}

export interface ComplianceCampaign {
  campaign_id: string;
  programme_id: string;
  stages: Record<NotificationStage, NotificationApplication>;
  created_at: string;
  cod_achieved: boolean;
  cod_date: string | null;
}

export interface StageSummary {
  stage: NotificationStage;
  total_tests: number;
  compliant: number;
  non_compliant: number;
  pending: number;
  conditional: number;
  submitted_at: string | null;
  approved_at: string | null;
  overall_status: string;
}
